### ISSUE TITLE
a backup of older checkpoint.pth?

### ISSUE THREAD

**Issue by bloodraven66** [espnet/espnet#5407]
**Describe the bug**
While training an ASR model, the machine crashed while saving checkpoint.pth, and it got corrupted. Due to this, I was unable to resume training. The *epoch.pth files do not have reporter, optimizers, schedulers, scaler, so there is no way to use a previous epoch checkpoint to resume training while retaining the keys. Can we maintain an older version of checkpoint.pth, so that in these situations, we can resume the training as usual with the backup checkpoint.pth?

**Basic environments:**
 - OS information: Linux 5.15.0-78-generic # 85~20.04.1-Ubuntu SMP Mon Jul 17 09:42:39 UTC 2023 x86_64
 - python version: [e.g. 3.7.3 (default, Mar 27 2019, 22:11:17) [GCC 7.3.0]]
 - espnet version: 202211
 - pytorch version: 1.12.1+cu116

**Task information:**
 - Task: ASR
 - Recipe: librispeech_100
 - ESPnet2

**To Reproduce**
1. move to a recipe directory, e.g., `cd egs/librispeech/asr1`
2. execute `run.sh` with specific arguments, e.g., `run.sh --stage 3 --ngp 1`
3. interrupt the task while saving checkpoint.pth

**Comment by sw005320** [comment]
@pyf98, can you tell me how you handle resuming your training?
Do you have the same experience?

**Comment by kamo-naoyuki** [comment]
It is impossible for now.  Indeed, I dropped the function when I implemented espnet2's trainer because I didn't expect someone wants it. 

As you said, now we have two kinds of pth to be saved when training:

- checkpoint.pth: Having all information about model, scheduler, optimizer, and etc for the latest epoch.
- *epoch.pth: The information of model for each epoch.

I actually *epoch.pth can have all information as same as checkpoint.pth and there are no strong motivations for the current style. I preferred the current style because we expect *epoch.pth is distributed as the pre-trained model. Threfore, we should keep it smaller size, and I decided to drop the other information from *epoch.pth.

The easiest way to implement your idea is saving checkpoint.pth for each epcoh, e.g.  `*checkpoint_*epoch.pth`. If you are motivated, please implement.

**Comment by bloodraven66** [comment]
> It is impossible for now. Indeed, I dropped the function when I implemented espnet2's trainer because I didn't expect someone wants it.
> 
> As you said, now we have two kinds of pth to be saved when training:
> 
> * checkpoint.pth: Having all information about model, scheduler, optimizer, and etc for the latest epoch.
> * *epoch.pth: The information of model for each epoch.
> 
> I actually *epoch.pth can have all information as same as checkpoint.pth and there are no strong motivations for the current style. I preferred the current style because we expect *epoch.pth is distributed as the pre-trained model. Threfore, we should keep it smaller size, and I decided to drop the other information from *epoch.pth.
> 
> The easiest way to implement your idea is saving checkpoint.pth for each epcoh, e.g. `*checkpoint_*epoch.pth`. If you are motivated, please implement.

I see.
I have added the following changes locally (in espnet2/train.trainer.py) as shown below, before saving checkpoint.pth
```
import os, shutil
if os.path.exists(output_dir / "checkpoint.pth"):
  shutil.copyfile(output_dir / "checkpoint.pth", output_dir / "checkpoint_prev.pth")
```

**Comment by pyf98** [comment]
> @pyf98, can you tell me how you handle resuming your training? Do you have the same experience?

Hi, I just resumed from the previous `checkpoint.pth` since it was always successfully saved without error. If an error happens during saving the checkpoint, then it is indeed difficult to resume.
